pkg/ndm-grpc/server/services/listBlockDevice.go

```diff
@@ -0,0 +1,192 @@
+
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/openebs/node-disk-manager/cmd/ndm_daemonset/controller"
+
+	"github.com/openebs/node-disk-manager/pkg/hierarchy"
+	"github.com/openebs/node-disk-manager/pkg/ndm-grpc/server"
+	"github.com/sirupsen/logrus"
+)
+
+// Node helps in using types defined in Server
+type Node struct {
+	server.Node
+}
+
+// NewNode is a constructor
+func NewNode(l *logrus.Logger) *Node {
+	return &Node{server.Node{Log: l}}
+}
+
+type disks []protos.BlockDevice
+
+// ListBlockDevices gives the status of iSCSI service
+func (n *Node) ListBlockDevices(ctx context.Context, null *protos.Null) (*protos.BlockDevices, error) {
+
+	n.Log.Info("Listing block devices")
+
+	ctrl, err := controller.NewController()
+	if err != nil {
+		n.Log.Errorf("Error creating a controller %v", err)
+		return nil, status.Errorf(codes.NotFound, "Namespace not found")
+	}
+
+	err = ctrl.SetControllerOptions(controller.NDMOptions{ConfigFilePath: "/host/node-disk-manager.config"})
+	if err != nil {
+		n.Log.Errorf("Error setting config to controller %v", err)
+		return nil, status.Errorf(codes.Internal, "Error setting config to controller")
+	}
+
+	blockDeviceList, err := ctrl.ListBlockDeviceResource(false)
+	if err != nil {
+		n.Log.Errorf("Error listing block devices %v", err)
+		return nil, status.Errorf(codes.Internal, "Error fetching list of disks")
+	}
+
+	if len(blockDeviceList.Items) == 0 {
+		n.Log.Info("No items found")
+	}
+
+	blockDevices := make([]*protos.BlockDevice, 0)
+
+	parentNames, err := GetParentDisks(n, blockDeviceList)
+	if err != nil {
+		n.Log.Errorf("Error fetching Parent disks %v", err)
+	}
+
+	// Commenting this section for now as Holders and Slaves do not yet support lsblk types
+	// holderNames, err := GetHolders(n, blockDeviceList)
+	// if err != nil {
+	// 	n.Log.Errorf("Error fetching Holders %v", err)
+	// }
+
+	// slaveNames, err := GetSlaves(n, blockDeviceList)
+	// if err != nil {
+	// 	n.Log.Errorf("Error fetching slaves %v", err)
+	// }
+
+	// for _, name := range holderNames {
+
+	// 	blockDevices = append(blockDevices, &protos.BlockDevice{
+	// 		Name: name,
+	// 		Type: "Holder",
+	// 	})
+	// }
+
+	// for _, name := range slaveNames {
+
+	// 	blockDevices = append(blockDevices, &protos.BlockDevice{
+	// 		Name: name,
+	// 		Type: "Slaves",
+	// 	})
+	// }
+
+	for _, name := range parentNames {
+
+		blockDevices = append(blockDevices, &protos.BlockDevice{
+			Name:       name,
+			Type:       "Disk",
+			Partitions: GetPartitions(n, name),
+		})
+	}
+
+	return &protos.BlockDevices{
+		Blockdevices: blockDevices,
+	}, nil
+}
+
+// GetHolders gets the holders of a particular device
+func GetHolders(n *Node, BL *v1alpha1.BlockDeviceList) ([]string, error) {
+
+	blockDeviceNames := make([]string, 0)
+	for _, bd := range BL.Items {
+		device := hierarchy.Device{Path: bd.Spec.Path}
+		depDevices, err := device.GetDependents()
```

Comment:
try limiting the call to `device.GetDependents`, It actually parses sysfs.

For each device in list you can call `GetDependents` only once and then get all the dependent devices, which can be further updated into the `Node` to get the required result.